probe/endpoint/procspy/background_reader_linux.go

```diff
@@ -0,0 +1,144 @@
+package procspy
+
+import (
+	"bytes"
+	"io"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/weaveworks/scope/probe/process"
+)
+
+const (
+	initialRateLimitPeriod = 50 * time.Millisecond  // Read 20 * fdBlockSize file descriptors (/proc/PID/fd/*) per namespace per second
+	maxRateLimitPeriod     = 500 * time.Millisecond // Read at least 2 * fdBlockSize file descriptors per namespace per second
+	minRateLimitPeriod     = initialRateLimitPeriod
+	fdBlockSize            = uint64(300) // Maximum number of /proc/PID/fd/* files to stat per rate-limit period
+	// (as a rule of thumb going through each block should be more expensive than reading /proc/PID/tcp{,6})
+	targetWalkTime = 10 * time.Second // Aim at walking all files in 10 seconds
+)
+
+type backgroundReader struct {
+	stopc         chan struct{}
+	mtx           sync.Mutex
+	latestBuf     *bytes.Buffer
+	latestSockets map[uint64]*Proc
+}
+
+// starts a rate-limited background goroutine to read the expensive files from
+// proc.
+func newBackgroundReader(walker process.Walker) *backgroundReader {
+	br := &backgroundReader{
+		latestBuf: bytes.NewBuffer(make([]byte, 0, 5000)),
```

Comment:
Also, you will probably want to allocate an empty `latestSockets: map[uint64]*Proc{}` here, so that someone who calls e.g. `getWalkedProcPid` before the first run doesn't get surprised by a nil map.